### PR TITLE
chore: parse the diff of two YAML files for properties with same names but different kinds

### PR DESCRIPTION
**Pull request by Lou1415926**
Address https://github.com/aws/copilot-cli/pull/4576#discussion_r1125105443

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

**Review by efekarakus — APPROVED** [review]
:shipit:

File: internal/pkg/template/diff/diff_test.go [hunk 1]
```diff
@@ -127,6 +127,96 @@ func TestFrom_Parse(t *testing.T) {
...
+							key: "Mary",
+							children: map[string]*Node{
+								"Dialogue": leafDialogue,
+							},
+						},
+					},
+				}
+			},
+		},
+		"change a map to list": {
```
Comment (huanjani) [inline internal/pkg/template/diff/diff_test.go:186]: The mock text is AMAZING. ❤️

File: internal/pkg/template/diff/diff.go [hunk 2]
```diff
@@ -42,11 +42,12 @@ func (from From) Parse(to []byte) (*Node, error) {
 }
 
 func parse(from, to *yaml.Node, key string) (*Node, error) {
-	if to == nil || from == nil {
+	// Handle base cases.
+	if (to == nil || from == nil) || to.Kind != from.Kind {
```
Comment (iamhopaul123) [inline internal/pkg/template/diff/diff.go]:
```suggestion
	if to == nil || from == nil || to.Kind != from.Kind {
```

File: internal/pkg/template/diff/diff_test.go [hunk 3]
```diff
@@ -127,6 +127,96 @@ func TestFrom_Parse(t *testing.T) {
...
+  Dialogue:
+    - Bear: "I know I'm supposed to keep an eye on you"
+      Tone: disappointed
+    - Dog: "ikr"
+      Tone: pleased`,
+			old: `Mary:
+  Dialogue:
+    Bear: (disappointed) "I know I'm supposed to keep an eye on you"
+    Dog: (pleased) "ikr"`,
+			wanted: func() *Node {
```
Comment (iamhopaul123) [inline internal/pkg/template/diff/diff_test.go:197]: Why do we use `func()`?

**Review by iamhopaul123 — APPROVED** [review]
LGTM just one nit and a question

File: internal/pkg/template/diff/diff_test.go (hunk 3 above)
Comment (Lou1415926) [reply internal/pkg/template/diff/diff_test.go:197]: Without `func()` it'd have to be super deeply nested, whereas now I can define `leaf` var and reduce the nesting. It's just a matter of readability.

**Pull request merged**